### DISCUSSION TITLE
Why index is not applied?

### DISCUSSION THREAD

**Discussion by vitaliishandra** [JanusGraph/janusgraph#3619] [Q&A] [answered]
We have index created with 'byAccountId2' with 'Enabled' status:
![image](https://user-images.githubusercontent.com/56965286/221165080-79bc4f96-5035-4c64-b743-c869a58568a0.png)
It works great when for example when we query the whole graph:
![image](https://user-images.githubusercontent.com/56965286/221165412-a811a781-737b-4dc8-84ab-3acf9f678f6c.png)
However for such query it didn't work:
![image](https://user-images.githubusercontent.com/56965286/221166073-ea85a830-6296-4c17-b60a-26c4ce87016a.png)
![image](https://user-images.githubusercontent.com/56965286/221166214-98962e96-2b67-45bc-9ac4-937cffce8af9.png)
The in step returns 300k vertexes.
If I remove filter by AccountId, everything is ok and query is executed immediately. 
I tried to run REINDEX but it didn't help.
Any ideas how what can be wrong?

**Comment by porunov** [comment, marked as answer]
In your second query you are filtering by `uid`, but your `byUid` index is disabled. You need to enable index first before it can be used.

Also, graph indices are not applied after you are traversing from the initial filtered result.

Vertex centric indices can be used to filter connected edges in traversals when necessary, but in your case it won't help because you are filtering by adjacent vertices and not edges.

To filter result by adjacent vertex properties you may be interested in the next feature, but it's not implemented yet: https://github.com/JanusGraph/janusgraph/discussions/2287

**Reply by vitaliishandra** [reply]
Thank you for you answer. Regarding byUid - it is everything OK with it, as we have another index ENABLED byUniqueId. 
Is it possible to somehow rewrite this query to apply index as now it took too long, beacuse first part returns ~300k vertexes: g.V().has('uid', within(...)).in("Direct", "Indirect") ?

**Reply by vitaliishandra** [reply]
@porunov Will vertex centric index will solve my issue in case I will move/duplciate property AccountId to Edge property: g.V().has('uid', within(...)).inE("Direct", "Indirect").has('AccountId', 257).out() ?

**Reply by porunov** [reply]
> @porunov Will vertex centric index will solve my issue in case I will move/duplciate property AccountId to Edge property: g.V().has('uid', within(...)).inE("Direct", "Indirect").has('AccountId', 257).out() ?

Yes, in such case vertex centric index will be applied to your query
